### ISSUE TITLE
Allow comparison patterns to declare a variable

### ISSUE THREAD

**Issue by YairHalberstadt** [dotnet/csharplang#3709]
eg.
```csharp
public class C
{
    public void CheckLength(string str) {
        if (str.Length is > 100 length)
            throw new Exception($"{length} is greater than maxlength of 100");
    }
}
```

**Comment by YairHalberstadt** [comment]
Current workaround:

```csharp
public class C
{
    public void CheckLength(string str) {
        if (str.Length is var length and > 100)
            throw new Exception($"{length} is greater than maxlength of 100");
    }
}
```

**Comment by jnm2** [comment]
I've been writing as `str.Length is > 100 and var length` because it's closer. I'd also like `str.IndexOf(c) is not -1 index`.

**Comment by 333fred** [comment]
You can use `and` patterns:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDZICYgNQA+AAgEwCMAsAFDVEDMABKQwMLUDe1D3TjRALKwAWcAMYBrADJwAdgHMYQgBREyABgawEASgacqPQwzQAzBkq0A6afMXGoDAHwN1GiDOwMAbhAQMMsgpC2lxGYYoIAPYA7gwycLEAogAeonAADjBokTJKACQAROwBtkIAvvYMcghwEPB+iu4MkMklQQyRZq4F2gDcodxl1GVAA=

**Comment by spydacarnage** [comment]
This might just be a bad example for the requested feature, because why would you not just use:
```csharp
public class C
{
    public void CheckLength(string str) {
        if (str.Length > 100)
            throw new Exception($"{str} is greater than maxlength of 100");
    }
}
```

**Comment by YairHalberstadt** [comment]
It was purely a made up on the spot example. Let's say that I don't won't very long exception messages :-)

**Comment by spydacarnage** [comment]
Fair enough :-)

**Comment by TonyValenti** [comment]
> eg.
> 
> ```cs
> public class C
> {
>     public void CheckLength(string str) {
>         if (str.Length is > 100 length)
>             throw new Exception($"{length} is greater than maxlength of 100");
>     }
> }
> ```

I am a huge fan of this.

**Comment by TonyValenti** [comment]
I would use it for declaring variables in my IF statements that are the result of a computation.  Basically like the following:

````
public class C
{
    public void Test1(int a, int b) {
        if(a + b is > 100 c && c <200){
                throw new Exception($@"{c} is not in range");
        }

    }
}
````

**Comment by DavidArno** [comment]
@TonyValenti,

Your example highlights my concern with this proposal. Your example can already be written as:
```cs
if (a + b is var c and > 100 and < 200)
{
        throw new Exception($@"{c} is not in range");
}
```
which takes the "pure pattern" approach rather than mixing patterns and boolean expressions as your example does. So this proposal risks making patterns more complex just to save having to type `and var ` as per @YairHalberstadt /@jnm2's examples. I see their examples as highlighting how to use patterns rather than in any way being a "work-around".

**Comment by canton7** [comment]
Agreed. We consistently emphasise that readability trumps character count.

`str.Length is > 100 length` reads to me as "(str.Length) (is greater than) (100 length)", and leaves me wondering what on earth "100 length" is.  I know you intend it to be parsed as "((str.Length) (is greater than) (100)) (length)", but it takes me a couple of scans through it to get to that point.

At least `str.Length is var length and > 100`, although still hard to parse, separates the "variable declaring" bit and the condition bit.

That said, I'd still write this as:

```
int length = str.Length;
if (length > 100) { }
```

**Comment by canton7** [comment]
These are also just weird:

```cs
if (str.Length is > 50 length and < 100)
if (str.Length is > 50 and < 100 length)
```

**Comment by Flutterish** [comment]
Id much rather just write
```cs
if ( ( var length = str.Length ) > 100 )
```
Ofc not correct syntax as of now
